// @apollo/client/link/http/selectHttpOptionsAndBody.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var tslib_es6 = require('../../../../tslib/tslib.es6.js');
require('../../../../graphql/version.mjs.js');
require('../../../../graphql/language/ast.mjs.js');
require('../../../../graphql/language/directiveLocation.mjs.js');
require('../../../../graphql/language/kinds.mjs.js');
require('../../../../graphql/language/tokenKind.mjs.js');
require('../../../../graphql/jsutils/instanceOf.mjs.js');
var printer = require('../../../../graphql/language/printer.mjs.js');
require('../../../../graphql/type/directives.mjs.js');
require('../../../../graphql/type/introspection.mjs.js');
require('../../../../graphql/language/visitor.mjs.js');
require('../../../../graphql/validation/specifiedRules.mjs.js');
require('../../../../graphql/type/scalars.mjs.js');
require('../../../../graphql/validation/rules/KnownTypeNamesRule.mjs.js');
require('../../../../graphql/validation/rules/PossibleTypeExtensionsRule.mjs.js');
require('../../../../graphql/utilities/extendSchema.mjs.js');
require('../../../../graphql/utilities/findBreakingChanges.mjs.js');

var defaultHttpOptions = {
    includeQuery: true,
    includeExtensions: false,
};
var defaultHeaders = {
    accept: '*/*',
    'content-type': 'application/json',
};
var defaultOptions = {
    method: 'POST',
};
var fallbackHttpConfig = {
    http: defaultHttpOptions,
    headers: defaultHeaders,
    options: defaultOptions,
};
var defaultPrinter = function (ast, printer) { return printer(ast); };
function selectHttpOptionsAndBodyInternal(operation, printer$1) {
    var configs = [];
    for (var _i = 2; _i < arguments.length; _i++) {
        configs[_i - 2] = arguments[_i];
    }
    var options = {};
    var http = {};
    configs.forEach(function (config) {
        options = tslib_es6.__assign(tslib_es6.__assign(tslib_es6.__assign({}, options), config.options), { headers: tslib_es6.__assign(tslib_es6.__assign({}, options.headers), headersToLowerCase(config.headers)) });
        if (config.credentials) {
            options.credentials = config.credentials;
        }
        http = tslib_es6.__assign(tslib_es6.__assign({}, http), config.http);
    });
    var operationName = operation.operationName, extensions = operation.extensions, variables = operation.variables, query = operation.query;
    var body = { operationName: operationName, variables: variables };
    if (http.includeExtensions)
        body.extensions = extensions;
    if (http.includeQuery)
        body.query = printer$1(query, printer.print);
    return {
        options: options,
        body: body,
    };
}
function headersToLowerCase(headers) {
    if (headers) {
        var normalized_1 = Object.create(null);
        Object.keys(Object(headers)).forEach(function (name) {
            normalized_1[name.toLowerCase()] = headers[name];
        });
        return normalized_1;
    }
    return headers;
}

exports.defaultPrinter = defaultPrinter;
exports.fallbackHttpConfig = fallbackHttpConfig;
exports.selectHttpOptionsAndBodyInternal = selectHttpOptionsAndBodyInternal;
