// @apollo/client/core/index.js
import '../utilities/globals/index.js';
export { ApolloClient } from './ApolloClient.js';
export { ObservableQuery } from './ObservableQuery.js';
export { NetworkStatus } from './networkStatus.js';
export { ApolloError, isApolloError } from '../errors/index.js';
export { ApolloCache } from '../cache/core/cache.js';
import '../../../graphql/version.mjs.js';
import '../../../graphql/language/ast.mjs.js';
import '../../../graphql/language/directiveLocation.mjs.js';
import '../../../graphql/language/kinds.mjs.js';
import '../../../graphql/language/tokenKind.mjs.js';
import '../../../graphql/jsutils/instanceOf.mjs.js';
import '../../../graphql/language/visitor.mjs.js';
import '../../../graphql/type/directives.mjs.js';
import '../../../graphql/type/introspection.mjs.js';
import '../../../graphql/validation/specifiedRules.mjs.js';
import '../../../graphql/type/scalars.mjs.js';
import '../../../graphql/validation/rules/KnownTypeNamesRule.mjs.js';
import '../../../graphql/validation/rules/PossibleTypeExtensionsRule.mjs.js';
import '../../../graphql/utilities/extendSchema.mjs.js';
import '../../../graphql/utilities/findBreakingChanges.mjs.js';
export { isReference, makeReference } from '../utilities/graphql/storeUtils.js';
import '../utilities/graphql/transform.js';
import '../utilities/common/mergeDeep.js';
import '../utilities/observables/Observable.js';
import '../utilities/observables/Concast.js';
import '../utilities/common/canUse.js';
import '../cache/inmemory/entityStore.js';
export { InMemoryCache } from '../cache/inmemory/inMemoryCache.js';
export { makeVar } from '../cache/inmemory/reactiveVars.js';
import '../cache/inmemory/policies.js';
import '../cache/inmemory/object-canon.js';
import '../link/core/empty.js';
import '../link/core/from.js';
import '../link/core/split.js';
import '../link/core/concat.js';
export { execute } from '../link/core/execute.js';
export { ApolloLink } from '../link/core/ApolloLink.js';
export { createHttpLink } from '../link/http/createHttpLink.js';
export { HttpLink } from '../link/http/HttpLink.js';
import { setVerbosity } from '../../../ts-invariant/lib/invariant.js';
export { setVerbosity as setLogVerbosity } from '../../../ts-invariant/lib/invariant.js';
export { disableExperimentalFragmentVariables, disableFragmentWarnings, enableExperimentalFragmentVariables, gql, resetCaches } from '../../../graphql-tag/lib/index.js';
import DEV from '../utilities/globals/DEV.js';

setVerbosity(DEV ? "log" : "silent");
