// @apollo/client/utilities/graphql/directives.js
import '../globals/index.js';
import '../../../../graphql/version.mjs.js';
import '../../../../graphql/language/ast.mjs.js';
import '../../../../graphql/language/directiveLocation.mjs.js';
import '../../../../graphql/language/kinds.mjs.js';
import '../../../../graphql/language/tokenKind.mjs.js';
import '../../../../graphql/jsutils/instanceOf.mjs.js';
import { visit } from '../../../../graphql/language/visitor.mjs.js';
import '../../../../graphql/type/directives.mjs.js';
import '../../../../graphql/type/introspection.mjs.js';
import '../../../../graphql/validation/specifiedRules.mjs.js';
import '../../../../graphql/type/scalars.mjs.js';
import '../../../../graphql/validation/rules/KnownTypeNamesRule.mjs.js';
import '../../../../graphql/validation/rules/PossibleTypeExtensionsRule.mjs.js';
import '../../../../graphql/utilities/extendSchema.mjs.js';
import '../../../../graphql/utilities/findBreakingChanges.mjs.js';
import { invariant } from '../../../../ts-invariant/lib/invariant.js';

function shouldInclude(_a, variables) {
    var directives = _a.directives;
    if (!directives || !directives.length) {
        return true;
    }
    return getInclusionDirectives(directives).every(function (_a) {
        var directive = _a.directive, ifArgument = _a.ifArgument;
        var evaledValue = false;
        if (ifArgument.value.kind === 'Variable') {
            evaledValue = variables && variables[ifArgument.value.name.value];
            __DEV__ ? invariant(evaledValue !== void 0, "Invalid variable referenced in @".concat(directive.name.value, " directive.")) : invariant(evaledValue !== void 0, 37);
        }
        else {
            evaledValue = ifArgument.value.value;
        }
        return directive.name.value === 'skip' ? !evaledValue : evaledValue;
    });
}
function getDirectiveNames(root) {
    var names = [];
    visit(root, {
        Directive: function (node) {
            names.push(node.name.value);
        },
    });
    return names;
}
function hasDirectives(names, root) {
    return getDirectiveNames(root).some(function (name) { return names.indexOf(name) > -1; });
}
function hasClientExports(document) {
    return (document &&
        hasDirectives(['client'], document) &&
        hasDirectives(['export'], document));
}
function isInclusionDirective(_a) {
    var value = _a.name.value;
    return value === 'skip' || value === 'include';
}
function getInclusionDirectives(directives) {
    var result = [];
    if (directives && directives.length) {
        directives.forEach(function (directive) {
            if (!isInclusionDirective(directive))
                return;
            var directiveArguments = directive.arguments;
            var directiveName = directive.name.value;
            __DEV__ ? invariant(directiveArguments && directiveArguments.length === 1, "Incorrect number of arguments for the @".concat(directiveName, " directive.")) : invariant(directiveArguments && directiveArguments.length === 1, 38);
            var ifArgument = directiveArguments[0];
            __DEV__ ? invariant(ifArgument.name && ifArgument.name.value === 'if', "Invalid argument for the @".concat(directiveName, " directive.")) : invariant(ifArgument.name && ifArgument.name.value === 'if', 39);
            var ifValue = ifArgument.value;
            __DEV__ ? invariant(ifValue &&
                (ifValue.kind === 'Variable' || ifValue.kind === 'BooleanValue'), "Argument for the @".concat(directiveName, " directive must be a variable or a boolean value.")) : invariant(ifValue &&
                (ifValue.kind === 'Variable' || ifValue.kind === 'BooleanValue'), 40);
            result.push({ directive: directive, ifArgument: ifArgument });
        });
    }
    return result;
}

export { getDirectiveNames, getInclusionDirectives, hasClientExports, hasDirectives, shouldInclude };
