// @apollo/client/link/http/selectHttpOptionsAndBody.js
import { __assign } from '../../../../tslib/tslib.es6.js';
import '../../../../graphql/version.mjs.js';
import '../../../../graphql/language/ast.mjs.js';
import '../../../../graphql/language/directiveLocation.mjs.js';
import '../../../../graphql/language/kinds.mjs.js';
import '../../../../graphql/language/tokenKind.mjs.js';
import '../../../../graphql/jsutils/instanceOf.mjs.js';
import { print } from '../../../../graphql/language/printer.mjs.js';
import '../../../../graphql/type/directives.mjs.js';
import '../../../../graphql/type/introspection.mjs.js';
import '../../../../graphql/language/visitor.mjs.js';
import '../../../../graphql/validation/specifiedRules.mjs.js';
import '../../../../graphql/type/scalars.mjs.js';
import '../../../../graphql/validation/rules/KnownTypeNamesRule.mjs.js';
import '../../../../graphql/validation/rules/PossibleTypeExtensionsRule.mjs.js';
import '../../../../graphql/utilities/extendSchema.mjs.js';
import '../../../../graphql/utilities/findBreakingChanges.mjs.js';

var defaultHttpOptions = {
    includeQuery: true,
    includeExtensions: false,
};
var defaultHeaders = {
    accept: '*/*',
    'content-type': 'application/json',
};
var defaultOptions = {
    method: 'POST',
};
var fallbackHttpConfig = {
    http: defaultHttpOptions,
    headers: defaultHeaders,
    options: defaultOptions,
};
var defaultPrinter = function (ast, printer) { return printer(ast); };
function selectHttpOptionsAndBodyInternal(operation, printer) {
    var configs = [];
    for (var _i = 2; _i < arguments.length; _i++) {
        configs[_i - 2] = arguments[_i];
    }
    var options = {};
    var http = {};
    configs.forEach(function (config) {
        options = __assign(__assign(__assign({}, options), config.options), { headers: __assign(__assign({}, options.headers), headersToLowerCase(config.headers)) });
        if (config.credentials) {
            options.credentials = config.credentials;
        }
        http = __assign(__assign({}, http), config.http);
    });
    var operationName = operation.operationName, extensions = operation.extensions, variables = operation.variables, query = operation.query;
    var body = { operationName: operationName, variables: variables };
    if (http.includeExtensions)
        body.extensions = extensions;
    if (http.includeQuery)
        body.query = printer(query, print);
    return {
        options: options,
        body: body,
    };
}
function headersToLowerCase(headers) {
    if (headers) {
        var normalized_1 = Object.create(null);
        Object.keys(Object(headers)).forEach(function (name) {
            normalized_1[name.toLowerCase()] = headers[name];
        });
        return normalized_1;
    }
    return headers;
}

export { defaultPrinter, fallbackHttpConfig, selectHttpOptionsAndBodyInternal };
