// @mui/system/esm/getThemeValue.js
import borders from './borders.js';
import display from './display.js';
import flexbox from './flexbox.js';
import grid from './grid.js';
import positions from './positions.js';
import palette from './palette.js';
import boxShadow from './shadows.js';
import sizing from './sizing.js';
import spacing from './spacing.js';
import typography from './typography.js';

const filterPropsMapping = {
  borders: borders.filterProps,
  display: display.filterProps,
  flexbox: flexbox.filterProps,
  grid: grid.filterProps,
  positions: positions.filterProps,
  palette: palette.filterProps,
  shadows: boxShadow.filterProps,
  sizing: sizing.filterProps,
  spacing: spacing.filterProps,
  typography: typography.filterProps
};
const styleFunctionMapping = {
  borders,
  display,
  flexbox,
  grid,
  positions,
  palette,
  shadows: boxShadow,
  sizing,
  spacing,
  typography
};
const propToStyleFunction = Object.keys(filterPropsMapping).reduce((acc, styleFnName) => {
  filterPropsMapping[styleFnName].forEach(propName => {
    acc[propName] = styleFunctionMapping[styleFnName];
  });
  return acc;
}, {});

export { propToStyleFunction, styleFunctionMapping };
