// @mui/material/Backdrop/Backdrop.js
import _objectWithoutPropertiesLoose from '../../../@babel/runtime/helpers/esm/objectWithoutPropertiesLoose.js';
import _extends from '../../../@babel/runtime/helpers/esm/extends.js';
import * as React from 'react';
import PropTypes from 'prop-types';
import clsx from '../../../clsx/dist/clsx.m.js';
import '../../utils/esm/elementAcceptingRef.js';
import '../../utils/esm/elementTypeAcceptingRef.js';
import 'react-is';
import '../../utils/esm/ponyfillGlobal.js';
import '../../utils/esm/refType.js';
import '../../utils/esm/integerPropType.js';
import composeClasses from '../../utils/esm/composeClasses/composeClasses.js';
import '../../base/BadgeUnstyled/BadgeUnstyled.js';
import '../../base/BadgeUnstyled/badgeUnstyledClasses.js';
import '../../base/ButtonUnstyled/ButtonUnstyled.js';
import '../../base/ButtonUnstyled/buttonUnstyledClasses.js';
import '../../base/ClickAwayListener/ClickAwayListener.js';
import '../../base/FormControlUnstyled/FormControlUnstyled.js';
import '../../base/FormControlUnstyled/FormControlUnstyledContext.js';
import '../../base/FormControlUnstyled/formControlUnstyledClasses.js';
import '../../base/InputUnstyled/InputUnstyled.js';
import '../../base/InputUnstyled/inputUnstyledClasses.js';
import '../../base/ListboxUnstyled/useListbox.types.js';
import '../../base/MenuUnstyled/MenuUnstyled.js';
import '../../base/MenuUnstyled/MenuUnstyledContext.js';
import '../../base/MenuUnstyled/menuUnstyledClasses.js';
import '../../base/MenuItemUnstyled/MenuItemUnstyled.js';
import '../../base/MenuItemUnstyled/menuItemUnstyledClasses.js';
import '../../base/ModalUnstyled/ModalUnstyled.js';
import '../../base/ModalUnstyled/modalUnstyledClasses.js';
import '../../base/MultiSelectUnstyled/MultiSelectUnstyled.js';
import '../../base/NoSsr/NoSsr.js';
import '../../base/OptionGroupUnstyled/OptionGroupUnstyled.js';
import '../../base/OptionGroupUnstyled/optionGroupUnstyledClasses.js';
import '../../base/OptionUnstyled/OptionUnstyled.js';
import '../../base/OptionUnstyled/optionUnstyledClasses.js';
import '../../base/PopperUnstyled/PopperUnstyled.js';
import '../../base/Portal/Portal.js';
import '../../base/SelectUnstyled/SelectUnstyled.js';
import '../../base/SelectUnstyled/selectUnstyledClasses.js';
import '../../base/SliderUnstyled/SliderUnstyled.js';
import '../../base/SliderUnstyled/SliderValueLabelUnstyled.js';
import '../../base/SliderUnstyled/sliderUnstyledClasses.js';
import '../../base/SwitchUnstyled/SwitchUnstyled.js';
import '../../base/SwitchUnstyled/switchUnstyledClasses.js';
import '../../base/TabPanelUnstyled/TabPanelUnstyled.js';
import '../../base/TabPanelUnstyled/tabPanelUnstyledClasses.js';
import '../../base/TabsUnstyled/TabsUnstyled.js';
import '../../base/TabsUnstyled/TabsContext.js';
import '../../base/TabsUnstyled/tabsUnstyledClasses.js';
import '../../base/TabsListUnstyled/TabsListUnstyled.js';
import '../../base/TabsListUnstyled/tabsListUnstyledClasses.js';
import '../../base/TabUnstyled/TabUnstyled.js';
import '../../base/TabUnstyled/tabUnstyledClasses.js';
import '../../base/TextareaAutosize/TextareaAutosize.js';
import '../../base/TrapFocus/TrapFocus.js';
import styled from '../styles/styled.js';
import useThemeProps from '../styles/useThemeProps.js';
import Fade from '../Fade/Fade.js';
import { getBackdropUtilityClass } from './backdropClasses.js';
import { jsx } from 'react/jsx-runtime';

const _excluded = ["children", "component", "components", "componentsProps", "className", "invisible", "open", "transitionDuration", "TransitionComponent"];

const useUtilityClasses = ownerState => {
  const {
    classes,
    invisible
  } = ownerState;
  const slots = {
    root: ['root', invisible && 'invisible']
  };
  return composeClasses(slots, getBackdropUtilityClass, classes);
};

const BackdropRoot = styled('div', {
  name: 'MuiBackdrop',
  slot: 'Root',
  overridesResolver: (props, styles) => {
    const {
      ownerState
    } = props;
    return [styles.root, ownerState.invisible && styles.invisible];
  }
})(({
  ownerState
}) => _extends({
  position: 'fixed',
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'center',
  right: 0,
  bottom: 0,
  top: 0,
  left: 0,
  backgroundColor: 'rgba(0, 0, 0, 0.5)',
  WebkitTapHighlightColor: 'transparent'
}, ownerState.invisible && {
  backgroundColor: 'transparent'
}));
const Backdrop = /*#__PURE__*/React.forwardRef(function Backdrop(inProps, ref) {
  var _components$Root, _componentsProps$root;

  const props = useThemeProps({
    props: inProps,
    name: 'MuiBackdrop'
  });

  const {
    children,
    component = 'div',
    components = {},
    componentsProps = {},
    className,
    invisible = false,
    open,
    transitionDuration,
    // eslint-disable-next-line react/prop-types
    TransitionComponent = Fade
  } = props,
        other = _objectWithoutPropertiesLoose(props, _excluded);

  const ownerState = _extends({}, props, {
    component,
    invisible
  });

  const classes = useUtilityClasses(ownerState);
  return /*#__PURE__*/jsx(TransitionComponent, _extends({
    in: open,
    timeout: transitionDuration
  }, other, {
    children: /*#__PURE__*/jsx(BackdropRoot, {
      "aria-hidden": true,
      as: (_components$Root = components.Root) != null ? _components$Root : component,
      className: clsx(classes.root, className),
      ownerState: _extends({}, ownerState, (_componentsProps$root = componentsProps.root) == null ? void 0 : _componentsProps$root.ownerState),
      classes: classes,
      ref: ref,
      children: children
    })
  }));
});
process.env.NODE_ENV !== "production" ? Backdrop.propTypes
/* remove-proptypes */
= {
  // ----------------------------- Warning --------------------------------
  // | These PropTypes are generated from the TypeScript type definitions |
  // |     To update them edit the d.ts file and run "yarn proptypes"     |
  // ----------------------------------------------------------------------

  /**
   * The content of the component.
   */
  children: PropTypes.node,

  /**
   * Override or extend the styles applied to the component.
   */
  classes: PropTypes.object,

  /**
   * @ignore
   */
  className: PropTypes.string,

  /**
   * The component used for the root node.
   * Either a string to use a HTML element or a component.
   */
  component: PropTypes.elementType,

  /**
   * The components used for each slot inside the Backdrop.
   * Either a string to use a HTML element or a component.
   * @default {}
   */
  components: PropTypes.shape({
    Root: PropTypes.elementType
  }),

  /**
   * The props used for each slot inside the Backdrop.
   * @default {}
   */
  componentsProps: PropTypes.shape({
    root: PropTypes.object
  }),

  /**
   * If `true`, the backdrop is invisible.
   * It can be used when rendering a popover or a custom select component.
   * @default false
   */
  invisible: PropTypes.bool,

  /**
   * If `true`, the component is shown.
   */
  open: PropTypes.bool.isRequired,

  /**
   * The system prop that allows defining system overrides as well as additional CSS styles.
   */
  sx: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.oneOfType([PropTypes.func, PropTypes.object, PropTypes.bool])), PropTypes.func, PropTypes.object]),

  /**
   * The duration for the transition, in milliseconds.
   * You may specify a single timeout for all transitions, or individually with an object.
   */
  transitionDuration: PropTypes.oneOfType([PropTypes.number, PropTypes.shape({
    appear: PropTypes.number,
    enter: PropTypes.number,
    exit: PropTypes.number
  })])
} : void 0;

export { Backdrop as default };
